kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -112,65 +143,133 @@ else if (requestFilterResult.message() != null) {
+        if (throwable != null) {
+            nettyDrivenFuture.completeExceptionally(throwable);
+        }
+        else {
+            nettyDrivenFuture.complete(t);
+        }
+    }
+
+    private <T extends FilterResult> CompletableFuture<T> handleDeferredStage(ChannelHandlerContext ctx, CompletableFuture<T> stage,
+                                                                              DecodedFrame<?, ?> decodedFrame) {
+        if (!stage.isDone()) {
+            inboundChannel.config().setAutoRead(false);
+            var timeoutFuture = ctx.executor().schedule(() -> {
+                LOGGER.warn("{}: Filter {} was timed-out whilst processing {} {}", ctx.channel(), filterDescriptor(),
+                        decodedFrame instanceof DecodedRequestFrame ? "request" : "response", decodedFrame.apiKey());
+                stage.completeExceptionally(new TimeoutException("Filter %s was timed-out.".formatted(filterDescriptor())));
+            }, timeoutMs, TimeUnit.MILLISECONDS);
+            return stage.thenApply(filterResult -> {
+                timeoutFuture.cancel(false);
+                inboundChannel.config().setAutoRead(true);
+                var unused = writeFuture.whenComplete((u, t) -> {
+                    // Ensure any writes pending will be flushed upstream, towards the broker.
+                    ctx.flush();
+                });
+                return filterResult;
+            });
+        }
+        else {
+            return stage;
         }
     }
 
     @Override
     public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
-        if (msg instanceof DecodedResponseFrame<?> decodedFrame) {
-            if (decodedFrame instanceof InternalResponseFrame<?> frame && frame.isRecipient(filter)) {
+        if (msg instanceof InternalResponseFrame<?> decodedFrame) {
+            // jump the queue, let extra requests flow back to their sender
+            if (decodedFrame.isRecipient(filter)) {
                 if (LOGGER.isDebugEnabled()) {
                     LOGGER.debug("{}: Completing {} response for request sent by this filter{}: {}",
                             ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
                 }
-                CompletableFuture<ApiMessage> p = frame.promise();
+                CompletableFuture<ApiMessage> p = decodedFrame.promise();
                 p.complete(decodedFrame.body());
             }
             else {
-                var filterContext = new DefaultFilterContext(filter, ctx, decodedFrame, null, timeoutMs, sniHostname);
-                if (LOGGER.isDebugEnabled()) {
-                    LOGGER.debug("{}: Dispatching upstream {} response to filter {}: {}",
-                            ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
+                doRead(ctx, msg);
+            }
+        }
+        else if (readFuture.isDone()) {
```

Comment:
Should we care about successful vs unsuccessful completion here?

Reply:
At the moment, I don't think we need to care. By my reading of the FilterHander, those future will only ever be completed successfully, even in the cases where the connection is closed,  there is a filter ends exceptionally or a filter unexpected returns null.

Maybe something in motivate us to get more sophisticated later, but I think what we have is already a valuable step forward.